src/gluonnlp/data/datasetloader.py

```diff
@@ -0,0 +1,419 @@
+        self._sent_idx = 0
+        self._data_buffer = {}
+
+        self._dataset = [dataset[i] for i in iter(file_sampler)]
+        self._num_datasets = len(self._dataset)
+
+        # construct cached list
+        self._cached_dataset = []
+
+        # pre-fetch
+        for _ in range(self._prefetch):
+            self._push_next_dataset()
+
+    def _push_next_dataset(self):
+        """Assign next dataset workload to workers."""
+        current_dataset_idx = self._sent_idx * self._circle_length
+        if current_dataset_idx < self._num_datasets:
+            circle_length = min(self._circle_length,
+                                self._num_datasets - current_dataset_idx)
+            if self._circle_length > 1:
+                url = [self._dataset[current_dataset_idx + i] for i in range(circle_length)]
+            else:
+                url = self._dataset[current_dataset_idx]
+        else:
+            return
+        # push to worker asynchronously
+        async_ret = self._worker_pool.apply_async(
+            self._worker_fn, (url, self._dataset_fn, self._batch_sampler_fn))
+        # data buffer stores the async result
+        self._data_buffer[self._sent_idx] = async_ret
+        self._sent_idx += 1
+
+    def _next_dataset(self):
+        """Retrieve the next dataset. Returns None if no dataset is available."""
+        if self._rcvd_idx == self._sent_idx:
+            assert not self._data_buffer, 'Data buffer should be empty at this moment'
+            return None
+
+        assert self._rcvd_idx < self._sent_idx, \
+               'rcvd_idx must be smaller than sent_idx'
+        assert self._rcvd_idx in self._data_buffer, \
+               'fatal error with _next_dataset, rcvd_idx missing'
+
+        if len(self._cached_dataset) == 0 or self._data_buffer[self._rcvd_idx].ready():
+            ret = self._data_buffer.pop(self._rcvd_idx)
+            dataset, batch_sampler = ret.get()
+            if self._manager:
+                dataset = self._manager.ProxyArrayDataset(*dataset._data)
+            self._rcvd_idx += 1
+            if self._cached and len(self._cached_dataset) < self._num_max_cached:
+                self._cached_dataset.append((dataset, batch_sampler))
+        else:
+            dataset, batch_sampler = self._cached_dataset.pop(0)
+
+        return dataset, batch_sampler
+
+    def __next__(self):
+        """Next dataset"""
+        self._push_next_dataset()
+        result = self._next_dataset()
+
+        if result is None:
+            raise StopIteration
+
+        return result
+
+    def next(self):
+        """Next dataset"""
+        return self.__next__()
+
+    def __iter__(self):
+        """Returns the iterator object"""
+        return self
+
+
+def _manager_register():
+    BaseManager.register('ProxyArrayDataset', ProxyArrayDataset)
+
+
+class DatasetLoader:
+    """Loads data from a list of datasets and returns mini-batches of data.
+
+    One dataset is loaded at a time.
```

Comment:
How does this differ from the DatasetStream and prefetching logic?

Reply:
This one maintains two worker pools for loading & *preprocessing* files and generating mini-batches. Here, *preprocessing* means training samples are generated on-the-fly, and reloading the same file can generate different training samples. The preprocessing function needs to be provided by the user. One particular examples can be seen from BERT pretraining script. And, it does not use queue, instead, it uses shared object among processes. Basically, it implements original gluon.dataloader while allowing generating new samples on-the-fly.